### ISSUE TITLE
ZAP seems to incorrectly report path traversal vulnerability in Angular app

### ISSUE THREAD

**Issue by mikedouglasdev** [zaproxy/zaproxy#3735]
I'm running OWASP ZAP as part of an automated CI/CD process. I am doing a spider and active scan. The report showed that there is a Path Traversal vulnerability.

This is an Angular 2 site and the javascript application is downloaded and runs in the local web browser so there wouldn't be anything revealed on the server. When I view the URL in question with and without the "attack", the results are the same. This URL just downloads a JavaScript file to the browser and the querystring is ignored. We are using webpack to do the bundling.

Here are the two instances.  I'm not sure what it means by "etc" for evidence.  Is it just looking for that string inside the results?  This bundle has "etc" in the file 62 times, however all part of other words like "fetch" or "getCookie".

URL | https://mysite/js/vendor.ece5bf651436a14bea3e.bundle.js?query=c%3A%2F
Method | GET
Parameter | query
Attack | c:/
Evidence | etc
URL | https://mysite/js/main.e2b095063a05762f397f.bundle.js?query=c%3A%2F
Method | GET
Parameter | query
Attack | c:/
Evidence | etc

Please let me know if you need any other information.

Thanks,
Mike

**Comment by thc202** [comment]
Yes, seems to be a false positive. Does the file have also `bin` and `boot`? (The match is case sensitive.)

**Comment by mikedouglasdev** [comment]
It doesn't have them as whole word.  It only has them as part of other words.  bootstrap is in the file several times.  bin is only part of other words like bind.

**Comment by thc202** [comment]
Thanks! Yes, that leads to a false positive.

**Comment by claudijd** [comment]
I just ran into a similar false positive, running zap-full-scan.py in the dockerized strategy.

Here are the details:

URL: https://mysite/somejsfile.min.js?query=c%3A%2F
Method: GET
Parameter: query
Attack: c:/
Evidence: etc

It seems to incorrectly triggering on the following strings...

getComputedStyle
getCurrentTime
setContainer
getCurrentContainer
getCurrentPlayback
...

I believe the diagnosis here is that the regex match for path traversal is entirely too loose, which is causing these false positives.

**Comment by kingthorin** [comment]
Can you confirm that your responses also contain the strings "bin" and "boot" (without quotes of course)?

**Comment by claudijd** [comment]
@kingthorin yes, in my case, the following bin/boot string matches existed...

bind
tabindex
t.fn.bind
unbind
this.bind
this.onEvent.bind
r.bindEvents()
...(probably another ~100 matches on the substring "bin")...
e.prototype._bootstrap
this._bootstrap
this.bootstrapId
this._bootstrapAttempts

There were no spaces before/after any of these examples, as it was just returning the JS source.

**Comment by kingthorin** [comment]
@claudijd thanks. A fix is in the works :zap:

**Comment by Sandvasi** [comment]
Hey ,
Is this issue resolved ?
I am still seeing the same issue in the ZAP tool .
https://mysite.com/polyfills.dll.js?query=c%3A%2F


Parameter | query
Attack | c:/
Evidence | etc
Source : Active (6 - Path Traversal)

Thanks

**Comment by thc202** [comment]
Which version of Active scanner rules add-on?

**Comment by Sandvasi** [comment]
OWASP ZAP Version: 2.7.0
[id=ascanrules, version=28.0.0]

**Comment by thc202** [comment]
This should be fixed in the latest version (30), could you try with that one? [1]

[1] https://github.com/zaproxy/zap-core-help/wiki/HelpUiDialogsManageaddons

**Comment by spizzat2** [comment]
I just looked at the current github version of TestPathTraversal.java, and lines 618 through 622 are still only checking that "bin", "boot", and "etc" exist in the response, and that /\betc/b/ is present. That means you can easily trigger this by: 
    1) Having a tabindex on any element on the page,
    2) Using Bootstrap, and 
    3) Including the word "etc" on your page [as one would do to truncate a long list]

You should also be able to trigger this by having a page that includes a very short list of words that begin with the letter "b" [ "binary", "booting",etc...]. I'm sure most fixes could be torn down by absurd edge cases, but my first example isn't terribly unlikely, is it?

At a minimum, the word boundary check should be applied to each of the triggers. Ideally, something more robust would be used, but that's obviously much more complicated, and I'm sure checking three times is already a performance hit.

**Comment by kingthorin** [comment]
We’re open to PRs.

**Comment by thc202** [comment]
The scanner is also checking if etc/bin/boot exist in the response beforehand, so simply having that in the response should not trigger it. It would be great if you could provide the response that's triggering the FP.

**Comment by spizzat2** [comment]
I'm not comfortable posting the HTML here. I'm trying to create a minimum viable test case, but I can only seem to trigger it when I run it against the entire site, despite it only triggering on one request. I have confirmed that the response that is triggering the code does not have links to /etc/, /boot/, or /bin/ on the file system, but the HTML does meet the constraints posted above ( has "tabindex" attribute, uses Bootstrap JavaScript library, and contains the word "etc").
I've also confirmed that swapping "etc" for "et cetera" will prevent the false-positive.

**Comment by spizzat2** [comment]
Ok, I ran a few tests and found a minimum viable test case. It only seemed to trigger when the request got redirected, so I just redirected to different pages that cause the issue. One of the test pages literally just contained the words "bin boot etc", which triggered the alert. The second page was a little more "real world", in that it contained the string `<input tabindex="0" type="hidden" name="booting" id="etc" />`, which also triggered the alert. I've included the request, as well as the two responses.
[path-traversal-request-header.txt](https://github.com/zaproxy/zaproxy/files/3138918/path-traversal-request-header.txt)
[path-traversal-response-header-binbootetc.txt](https://github.com/zaproxy/zaproxy/files/3138919/path-traversal-response-header-binbootetc.txt)
[path-traversal-response-input.txt](https://github.com/zaproxy/zaproxy/files/3138920/path-traversal-response-input.txt)

Fixing this by adding word-boundary checks to "bin" and "boot" really just kicks the can down the road, though. It would only fix the second test, not the first. 
Once the code is doing regular expressions to get word boundaries, it _could_ try to look for anchor tags. However, that could possibly make the test too lax, and I think it's better to have false positives than miss actual vulnerabilities.
